_posts/2019-10-23-gpt-2/gpt2.Rmd

```diff
@@ -0,0 +1,317 @@
+
+Many subsequent NLP models built on the Transformer, but -- depending on purpose -- either picked up the encoder stack only, or just the decoder stack.
+GPT-2 was trained to predict consecutive words in a sequence. It is thus a _language model_, a term resounding the conception that an algorithm which can predict future words and sentences somehow has to _understand_ language (and a lot more, we might add).
+As there is no input to be encoded (apart from an optional one-time prompt), all that is needed is the stack of decoders. 
+
+In our experiments, we'll be using the biggest as-yet released pretrained model, but this being a pretrained model our degrees of freedom are limited. We can, of course, condition on different input prompts. In addition, we can influence the sampling algorithm used.
+
+## Sampling options with GPT-2
+
+Whenever a new token is to be predicted, a _softmax_ is taken over the vocabulary ^[For an introduction to how softmax activation behaves, see [Winner takes all: A look at activations and cost functions](https://blogs.rstudio.com/tensorflow/posts/2018-10-11-activations-intro/).]. Directly taking the softmax output amounts to maximum likelihood estimation. In reality, however, always choosing the maximum likelihood estimate results in highly repetitive output. 
+
+A natural option seems to be using the softmax outputs as probabilities: Instead of just taking the _argmax_, we sample from the output distribution. Unfortunately, this procedure has negative ramifications of its own. In a big vocabulary, very improbable words together make up a substantial part of the probability mass; at every step of generation, there is thus a non-negligible probability that an improbable word may be chosen. This word will now exert great influence on what is chosen next. In that manner, highly improbable sequences can build up. 
+
+The task thus is to navigate between the Scylla of determinism and the Charybdis of weirdness. With the GPT-2 model presented below, we have three options:
+
+- vary the _temperature_ (parameter `temperature`),
+- vary `top_k`, the number of tokens considered, or
```

Comment:
```suggestion
- vary `top_k`, the number of tokens considered; or
```

Reply:
I switched to semicolon-delimeted, since two of the bullets have complex structure. 🤓

Reply:
haha 
love it :-)